Wip/shake queue

mpickering opened the pull request:
Putting this up here so we can comment about what needs to be tidied up.

mpickering committed:
Use stale information for hover and completions

This introduces a new function `useWithStaleFast` which returns with
stale information WITHOUT checking freshness like `use` and
`useWithStale`.

Greatly improve debug logging

All actions triggered by shakeRun now also pass an identifier which
means that the debug logging shows which actions are starting/finishing

We also distinguish between internal and external events. By default
external events are ones triggered by runAction and the debug output
is displayed to the user in command line and --lsp mode.

In order to see internal logging statements, there is a new flag called
--verbose which also prints out internal events such as file
modification flushes.

Cleaner variant using runAfter

Step 1: Do not run actions with shakeRun

Queue implementation, living, breathing

Use a priority queue to schedule shake actions.

Most user actions are answered immediately with a cache but also
spawn a shake action to check the cached value we consulted was up to
date.

pepeiborra committed:
ShakeSession and shakeRunGently

Currently we start a new Shake session for every interaction with the Shake
database, including type checking, hovers, code actions, completions, etc.
Since only one Shake session can ever exist, we abort the active session if any
in order to execute the new command in a responsive manner.

This is suboptimal in many, many ways:

- A hover in module M aborts the typechecking of module M, only to start over!
- Read-only commands (hover, code action, completion) need to typecheck all the
  modules! (or rather, ask Shake to check that the typechecks are current)
- There is no way to run non-interfering commands concurrently

This is an experiment inspired by the 'ShakeQueue' of @mpickering, and
the follow-up discussion in https://github.com/mpickering/ghcide/issues/7

We introduce the concept of the 'ShakeSession' as part of the IDE state.
The 'ShakeSession' is initialized by a call to 'shakeRun', and survives until
the next call to 'shakeRun'. It is important that the session is restarted as
soon as the filesystem changes, to ensure that the database is current.

The 'ShakeSession' enables a new command 'shakeRunGently', which appends work to
the existing 'ShakeSession'. This command can be called in parallel without any
restriction.

pepeiborra committed:
Simplify by assuming there is always a ShakeSession

pepeiborra committed:
Improved naming and docs

pepeiborra committed:
Define runActionSync on top of shakeEnqueue

shakeRun is not correct as it never returns anymore

pepeiborra committed:
Drive progress reporting from newSession

The previous approach reused the shakeProgress thread,  which doesn't work anymore as ShakeSession keeps the ShakeDatabase open until the next edit

pepeiborra committed:
Deterministic progress messages in tests

Dropping the 0.1s sleep to ensure that progress messages during tests are
deterministic

pepeiborra committed:
Make kick explicit

This is required for progress reporting to work, see notes in shakeRun

As to whether this is the right thing to do:

1. Less magic, more explicit
2. There's only 2 places where kick is actually used

pepeiborra committed:
apply Neil's feedback

pepeiborra committed:
avoid a deadlock when the enqueued action throws

pepeiborra committed:
Simplify runAction + comments

pepeiborra committed:
use a Barrier for clarity

A Barrier is a smaller abstraction than an MVar, and the next version of the extra package will come with a suitably small implementation:

https://github.com/ndmitchell/extra/commit/98c2a83585d2ca0a9d961dd241c4a967ef87866a

pepeiborra committed:
Log timings for code actions, hovers and completions

mpickering commented on src/Development/IDE/Core/OfInterest.hs:
```diff
@@ -79,4 +79,5 @@ modifyFilesOfInterest state f = do
     files <- modifyVar var $ pure . dupe . f
     logDebug (ideLogger state) $ "Set files of interest to: " <> T.pack (show $ HashSet.toList files)
-    void $ shakeRun state []
+    let das = map (\nfp -> mkDelayedAction "OfInterest" (GetSpanInfo, nfp) Debug (use GetSpanInfo nfp)) (HashSet.toList files)
```
Do we want this or not?

wz1000 replied:
What is the point supposed to be?

mpickering replied:
On this branch there isn't the parent typechecking logic so if you edit a file A which is depended on by a file B then B won't get recompiled if you make a change in A without this logic.

mpickering commented on src/Development/IDE/Import/DependencyInformation.hs:
```diff
@@ -100,7 +100,7 @@ insertPathId path id PathIdMap{..} =
 
 insertImport :: FilePathId -> Either ModuleParseError ModuleImports -> RawDependencyInformation -> RawDependencyInformation
-insertImport (FilePathId k) v rawDepInfo = rawDepInfo { rawImports = IntMap.insert k v (rawImports rawDepInfo) }
+insertImport (FilePathId f) v rawDepInfo = rawDepInfo { rawImports = IntMap.insert f v (rawImports rawDepInfo) }
```
revert

mpickering commented on src/Development/IDE/Plugin/CodeAction.hs:
```diff
@@ -74,11 +74,12 @@ codeAction lsp state (TextDocumentIdentifier uri) _range CodeActionContext{_diag
     let text = Rope.toText . (_text :: VirtualFile -> Rope.Rope) <$> contents
         mbFile = toNormalizedFilePath' <$> uriToFilePath uri
-    (ideOptions, parsedModule, join -> env) <- runAction state $
+    (ideOptions, parsedModule, join -> env) <- runAction "CodeAction" state $
```
CodeActions are still really slow on GHC because of this.

mpickering commented on src/Development/IDE/Plugin/Completions.hs:
```diff
@@ -73,29 +74,36 @@ getCompletionsLSP
+    -- We mask here as otherwise completions interact badly with
+    -- modification events which cause the thread to get continually
+    -- killed before the action can finish.
+    mask_ $ do
```
Revert.

mpickering commented on src/Development/IDE/Core/Shake.hs:
```diff
@@ -223,22 +243,212 @@ data IdeState = IdeState
     }
 
 
+data QPriority = QPriority { retries :: Int
```
Move all the ShakeQueue stuff into another module

mpickering commented on src/Development/IDE/Core/Shake.hs:
```diff
@@ -255,6 +465,8 @@ setValues :: IdeRule k v
 setValues state key file val = modifyVar_ state $ \vals -> do
     -- Force to make sure the old HashMap is not retained
+    --let size = HMap.size vals
+    --traceEventIO ("METRIC:" ++ show size)
```
remove

mpickering commented on src/Development/IDE/Core/Shake.hs:
```diff
@@ -301,13 +513,16 @@ shakeOpen :: IO LSP.LspId
           -> IO IdeState
 shakeOpen getLspId eventer logger debouncer shakeProfileDir (IdeReportProgress reportProgress) opts rules = do
     inProgress <- newVar HMap.empty
+    shakeAbort <- newMVar $ return ()
```
remove

mpickering commented on src/Development/IDE/Spans/AtPoint.hs:
```diff
@@ -129,26 +129,42 @@ locationsAtPoint getHieFile _ideOptions pos =
-        eqName n n' = nameOccName n == nameOccName n' && nameModule_maybe n == nameModule_maybe n'
-        setFileName f (RealSrcSpan span) = RealSrcSpan (span { srcSpanFile = mkFastString f })
-        setFileName _ span@(UnhelpfulSpan _) = span
+        getSpan (Named name) = nameToLocation getHieFile name
```
THis refactoring is from goto type definition patch which is still not merged.

mpickering commented:
Patch overall looks quite clean once Shake.hs is sorted out.

mpickering commented:
My plan here now is to rebase this onto @pepeiborra's shake enqueue patch and wait for that to land before making a PR.

pepeiborra committed:
Rename shakeRun to shakeRestart

The action returned by shakeRun now blocks until another call to shakeRun is made, which is a change in behaviour,. but all the current uses of shakeRun ignore this action.

Since the new behaviour is not useful, this change simplifies and updates the docs and name accordingly

pepeiborra committed:
delete runActionSync as it's just runAction

pepeiborra committed:
restart shake session on new component created

pepeiborra committed:
requeue pending actions on session restart

pepeiborra committed:
hlint

pepeiborra committed:
Bumped the delay from 5 to 6

mpickering committed:
WIP

mpickering committed:
Working now I think

mpickering committed:
fix tests

mpickering committed:
Add some better logging

wz1000 committed:
Cleanup useWithStaleFast

wz1000 committed:
delete ShakeQueue
